runtime/runtime/src/lib.rs

```diff
@@ -1679,63 +1778,105 @@ impl Runtime {
+                let batch_outcome = batch_result?;
+                if let (Some((account, access_key)), Some(signer_id), Some(public_key)) = (
+                    batch_outcome.final_state.as_ref(),
+                    batch_outcome.signer_id(),
+                    batch_outcome.public_key(),
+                ) {
+                    set_account(state_update, signer_id.clone(), &account);
+                    set_access_key(
                         state_update,
-                        apply_state,
-                        &validated_tx,
-                        &cost,
-                        &mut processing_state.stats,
-                    ) {
-                        Ok(outcome) => outcome,
-                        Err(err) => {
-                            Self::handle_invalid_transaction(
-                                err,
-                                &tx_hash,
-                                processing_state.protocol_version,
-                                "process_transaction error",
-                            )?;
-                            continue;
-                        }
-                    };
-                    if receipt.receiver_id() == validated_tx.signer_id() {
-                        processing_state.local_receipts.push_back(receipt);
-                    } else {
-                        receipt_sink.forward_or_buffer_receipt(
-                            receipt,
-                            apply_state,
-                            state_update,
-                            processing_state.epoch_info_provider,
-                        )?;
-                    }
-                    let compute = outcome_with_id.outcome.compute_usage;
-                    let compute =
-                        compute.expect("`process_transaction` must populate compute usage");
-                    total.add(outcome_with_id.outcome.gas_burnt, compute)?;
-                    if !ProtocolFeature::ComputeCosts.enabled(processing_state.protocol_version) {
-                        assert_eq!(total.compute, total.gas, "Compute usage must match burnt gas");
-                    }
-                    processing_state.outcomes.push(outcome_with_id);
+                        signer_id.clone(),
+                        public_key.clone(),
+                        &access_key,
+                    );
+                    let last_tx_hash = batch_outcome
+                        .processed_transactions
+                        .last()
+                        .map(|pt| pt.transaction.get_hash())
+                        .unwrap_or_default();
+                    state_update
+                        .commit(StateChangeCause::TransactionProcessing { tx_hash: last_tx_hash });
+                } else {
+                    debug_assert!(
+                        batch_outcome.processed_transactions.is_empty(),
+                        "must be an empty batch result"
+                    );
```

Comment:
Is it possible for us to have an empty batch?  Could we not filter them out when we construct the batches?  And then some of the `Option`s in BatchResult do not need to be `Option` anymore.

Reply:
Even though we can filter empty batches, there's no guarantee that at least one transaction would pass verification, so empty outcome would still be possible